### ISSUE TITLE
[Bug] System.ArgumentException: 'Handle must be valid. Parameter name: type. After accessing a DateTime Preference

### ISSUE THREAD

**Issue by tkemp2290** [xamarin/Essentials#1334]
### Description
I have an app that currently only downloads a bunch of data from an API server. I haven't gotten to persisting the data in the internal DB yet. However, one of the functions of the app will be that it will only do a download on start up of the app if the last download occurred more than hours ago. As such, I am storing the DateTime of the last successful in shared preferences using Xamarin.Essentials. Part of the download routine access a string collection stored in SharedPreferences which is a collection of Ids for which to do the download for.

The problem is that when I try and access that preference for the first time, when the code then tries to access the string collection preference in the download routine i get the above crash. Remove the line that gets the DateTime preference beforehand and no crash.

### Steps to Reproduce

1. Run app
2. Access Last Downloaded Date Preference and work out difference between that and now
3. Access StoreIds String Collection Preference

### Expected Behavior
Does not crash and app can perform download routine
### Actual Behavior
App crashes with above exception
### Basic Information
Xamarin.Native using MvvmCross, not Xamarin.Forms

- Version with issue: 1.5.3.2
- Last known good version:
- IDE: Visual Studio 2019 Professional (Win 10)
- Platform Target Frameworks: 
  - Android: 9
- Android Support Library Version: I think MvvmCross 6.4.2 installs 28.0.0.3
- Nuget Packages:
MvvmCross 6.4.2
- Affected Devices:
Android Emulator

**Comment by jamesmontemagno** [comment]
Do you have sample code that we can look at?

**Comment by tkemp2290** [comment]
@jamesmontemagno 

here is the SharedPrefs Service
` 
public class SharedPreferencesPlatformService : ISharedPreferencesPlatformService
    {
        private Activity _topActivity;
        private ISharedPreferences _sharedPrefs;
        private ISharedPreferencesEditor _prefsEditor;
        public SharedPreferencesPlatformService()
        {
            _topActivity = Mvx.IoCProvider.Resolve<IMvxAndroidCurrentTopActivity>().Activity;
            _sharedPrefs = PreferenceManager.GetDefaultSharedPreferences(_topActivity.ApplicationContext);
            _prefsEditor = _sharedPrefs.Edit();
        }

        public void RemoveSecureString(string key)
        {
            SecureStorage.Remove(key);
        }

        public bool RetreieveBoolPreference(string settingKey)
        {
            return _sharedPrefs.GetBoolean(settingKey, false);
        }

        public DateTime RetrieveDateTimePreference(string key)
        {
            return Preferences.Get("key", DateTime.MinValue);
        }

        public int RetrieveIntPreference(string settingKey)
        {
            return _sharedPrefs.GetInt(settingKey, 0);
        }

        public async Task<string> RetrieveSecureString(string key)
        {
            return await SecureStorage.GetAsync(key);
        }

        public ICollection<string> RetrieveStringCollectionPreference(string settingKey)
        {
            ICollection<string> values = _sharedPrefs.GetStringSet(settingKey, new List<string>());

            return values?.ToList();
        }

        public string RetrieveStringPreference(string key)
        {
            return _sharedPrefs.GetString(key, string.Empty);
        }

        public void SaveBooleanPreference(string key, bool value)
        {
            _prefsEditor.PutBoolean(key, value).Commit();
        }

        public void SaveDateTimePreference(string key, DateTime value)
        {
            Preferences.Set(key, value);
        }

        public void SaveIntPreference(string key, int value)
        {
            _prefsEditor.PutInt(key, value).Commit();
        }

        public async Task SaveSecureString(string key, string value)
        {
            await SecureStorage.SetAsync(key, value);
        }

        public void SaveStringCollectionPreference(string key, ICollection<string> value)
        {
            _prefsEditor.PutStringSet(key, value).Commit();
        }

        public void SaveStringPreference(string key, string value)
        {
            _prefsEditor.PutString(key, value).Commit();
        }
    }
`

This is the AppStart.cs
` public class AppStart : MvxAppStart
    {
        private readonly IMvxNavigationService _navService;
        private readonly IIniFileReaderService _iniReaderService;
        private readonly ISharedPreferencesPlatformService _sharedPrefs;
        private readonly IMvxMessenger _messenger;
        private readonly IAuthenticationService _authService;
        private readonly IDataDowloadService _dataService;

        public AppStart(
            IMvxApplication application,
            IMvxNavigationService navigationService,
            IIniFileReaderService iniReaderService,
            ISharedPreferencesPlatformService sharedPrefs,
            IMvxMessenger messenger,
            IAuthenticationService authService,
            IDataDowloadService dataService
            ) : base(application, navigationService)
        {
            _navService = navigationService;
            _iniReaderService = iniReaderService;
            _sharedPrefs = sharedPrefs;
            _messenger = messenger;
            _authService = authService;
            _dataService = dataService;
        }

        protected override async Task NavigateToFirstViewModel(object hint = null)
        {
            _messenger.Publish(new ChangeSplashTextMessage(this, "Retrieving Authorisation file"));
            IniFile iniFile = await _iniReaderService.RetrieveAndReadIniFile();
            StoreIniFileDetailsInSharedPreferences(iniFile);
            _messenger.Publish(new ChangeSplashTextMessage(this, "Authorising Install with Rascal Server"));
            //perform authorization call to retrieve Auth Token if needed
            await _authService.AuthoriseWithServerIfNecessary();
            //perform initial full download if necessary
            var downloadedData = await _dataService.PerformFullDownloadIfNecessary();
            await _navService.Navigate<MainViewModel>();
        }

        private void StoreIniFileDetailsInSharedPreferences(IniFile iniFile)
        {
            _sharedPrefs.SaveStringCollectionPreference(Constants.STORE_IDS_KEY, iniFile.StoreIds);
            _sharedPrefs.SaveIntPreference(Constants.USER_TYPE_KEY, (int)iniFile.UserType);
        }
    }`

This is the DataDownloadService
`public class DataDownloadService : IDataDowloadService
    {
        private readonly IDataDownloadWebService _downloadService;
        private readonly ISharedPreferencesPlatformService _sharedPrefs;

        public DataDownloadService(IDataDownloadWebService downloadService, ISharedPreferencesPlatformService sharedPreferences)
        {
            _downloadService = downloadService;
            _sharedPrefs = sharedPreferences;
        }

        public async Task<DownloadedDataModel> PerformFullDownloadIfNecessary()
        {
            DownloadedDataModel downloadedData = null;
            //check last download time to see if a full download is necessary
            var lastDownloadDate = _sharedPrefs.RetrieveDateTimePreference(Constants.LAST_DOWNLOAD_PERFORMED_DATE_KEY);
            var differenceBetweenLastDownloadAndNow = DateTime.Now - lastDownloadDate;

            if (differenceBetweenLastDownloadAndNow > TimeSpan.FromHours(8))
            {
                downloadedData = await _downloadService.PerformFullDataDownload();
            }
            return downloadedData;
        }
    }`

and this is the DownloadWebService
`public class DataDownloadWebService : IDataDownloadWebService
    {
        private readonly IMvxMessenger _messenger;
        private readonly ISharedPreferencesPlatformService _sharedPrefs;
        private readonly IStoreWebService _storeWebService;
        private readonly IProductsWebService _productsWebService;
        private readonly IStaticDataWebService _staticDataWebService;
        private readonly IStorePerformanceWebService _storePerformanceWebService;
        private readonly IDistributorsWebService _distributorsWebService;
        private readonly IMessagesWebService _messagesWebService;
        private readonly IMatchedProductsWebService _matchedProductsService;
        private readonly IExpectedGoodsInWebService _expectedGoodsInWebService;

        public DataDownloadWebService(
            IMvxMessenger messenger,
            ISharedPreferencesPlatformService sharedPrefs,
            IStoreWebService storeWebService,
            IProductsWebService productsWebService,
            IStaticDataWebService staticDataWebService,
            IStorePerformanceWebService storePerformanceWebService,
            IDistributorsWebService distributorsWebService,
            IMessagesWebService messagesWebService,
            IMatchedProductsWebService matchedProductsWebService,
            IExpectedGoodsInWebService expectedGoodsInWebService)
        {
            _messenger = messenger;
            _sharedPrefs = sharedPrefs;
            _storeWebService = storeWebService;
            _productsWebService = productsWebService;
            _staticDataWebService = staticDataWebService;
            _storePerformanceWebService = storePerformanceWebService;
            _distributorsWebService = distributorsWebService;
            _messagesWebService = messagesWebService;
            _matchedProductsService = matchedProductsWebService;
            _expectedGoodsInWebService = expectedGoodsInWebService;
        }
        public async Task<DownloadedDataModel> PerformFullDataDownload()
        {
            _messenger.Publish(new DownloadStartedMessage(this));
            _messenger.Publish(new ChangeSplashTextMessage(this, "Doing a data download"));

            DownloadedDataModel downloadedData = null;

            var storeIds = _sharedPrefs.RetrieveStringCollectionPreference(Constants.STORE_IDS_KEY);

            var storeModels = new List<StoreDataModel>();
            var productModels = new List<ProductDataModel>();
            var deliveryIssueModels = new List<DeliveryIssueTypeDataModel>();
            var wholesalerIssueModels = new List<WholesalerIssueTypeDataModel>();
            var storePerformanceDataModels = new List<StorePerformanceDataModel>();
            var distributorDataModels = new List<DistributorLiteDataModel>();
            var messageTriggerModels = new List<MessageTriggerDataModel>();
            var stocktakeTypeModels = new List<StocktakeTypeDataModel>();
            var matchedProductDataModels = new List<MatchProductDataModel>();
            var expectedGoodsInDataModels = new List<ExpectedGoodsInDataModel>();

            foreach (var storeId in storeIds)
            {
                var storeData = await DownloadStoreData(storeId);
                if (storeData != null)
                {
                    storeModels.Add(storeData);
                    var performanceData = await DownloadStorePerformanceDataForStore(storeId);
                    if (performanceData != null)
                        storePerformanceDataModels.Add(performanceData);
                    var storeDistributors = await DownloadDistributorsForStore(storeId);
                    if (storeDistributors != null)
                        distributorDataModels.AddRange(storeDistributors);
                    var messageTriggers = await DownloadMessageTriggersForStore(storeId);
                    if (messageTriggers != null)
                        messageTriggerModels.AddRange(messageTriggers);
                    var stocktakeTypes = storeData.Categories
                    .SelectMany(c => c.Subcategories)
                    .SelectMany(sc => sc.AllowedStocktakeTypes)
                    .GroupBy(y => y.Id + y.Description)
                    .Select(y => y.First())
                    .ToList();
                    if (stocktakeTypes != null)
                        stocktakeTypeModels.AddRange(stocktakeTypes);
                    var matchedProducts = await DownloadMatchedProductsForStore(storeId, storeData.Categories.SelectMany(c => c.Subcategories).ToList());
                    if (matchedProducts != null)
                        matchedProductDataModels.AddRange(matchedProducts);
                    var expectedGoods = await GetExpectedGoodsInForStore(storeId);
                }

            }
            if(storeModels.Count >= 1)
            {                
                var productData = await DownloadProductDataForStores(storeIds);
                if (productData != null)
                    productModels.AddRange(productData);
            }

            //static data
            var deliveryIssues = await DownloadDeliveryIssuesData();
            if (deliveryIssues != null)
                deliveryIssueModels.AddRange(deliveryIssues);

            var wholesalerIssues = await DownloadWholesalerIssueData();
            if (wholesalerIssues != null)
                wholesalerIssueModels.AddRange(wholesalerIssues);

            _messenger.Publish(new DownloadFinishedMessage(this));

            return downloadedData;
        }

        private async Task<List<ExpectedGoodsInDataModel>> GetExpectedGoodsInForStore(string storeId)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Deliveries For Store: {storeId}", 40));

            return await _expectedGoodsInWebService.GetExpectedGoodsInForStore(storeId);
        }

        private async Task<List<MatchProductDataModel>> DownloadMatchedProductsForStore(string storeId, List<SubCategoryDataModel> subcategories)
        {
            var matchedProducts = new List<MatchProductDataModel>();
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Matched Products For Store: {storeId}", 30));
            foreach(var subcategory in subcategories)
            {
                var products = await _matchedProductsService.GetMatchedProductForStoreSubcategory(storeId, subcategory.SubcategoryId.ToString());
                if (products != null)
                    matchedProducts.AddRange(products);
            }
            return matchedProducts;
        }

        private async Task<List<MessageTriggerDataModel>> DownloadMessageTriggersForStore(string storeId)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Message Triggers For Store: {storeId}", 25));
            return await _messagesWebService.GetMessagesForStore(storeId);
        }

        private async Task<List<DistributorLiteDataModel>> DownloadDistributorsForStore(string storeId)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Distributor Data For Store: {storeId}", 20));
            return await _distributorsWebService.GetDistributorsForStore(storeId);
        }

        private async Task<StorePerformanceDataModel> DownloadStorePerformanceDataForStore(string storeId)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Store Performance Data For Store: {storeId}", 15));
            return await _storePerformanceWebService.GetStorePerformanceDataForStore(storeId);
        }

        private async Task<List<WholesalerIssueTypeDataModel>> DownloadWholesalerIssueData()
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Static Data", 70));
            return await _staticDataWebService.GetWholesalerIssueTypes();
        }

        private async Task<List<DeliveryIssueTypeDataModel>> DownloadDeliveryIssuesData()
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Static Data", 60));
            return await _staticDataWebService.GetDeliveryIssueTypes();
        }

        private async Task<IEnumerable<ProductDataModel>> DownloadProductDataForStores(ICollection<string> storeIds)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Product Data", 50));            

            return await _productsWebService.GetProducts(storeIds);
            
        }

        private async Task<StoreDataModel> DownloadStoreData(string storeId)
        {
            _messenger.Publish(new ChangeDownloadStepTextMessage(this, $"Downloading Store Data for Store: {storeId}", 10));
            return await _storeWebService.GetStoreDetails(storeId);
        }
    }`

The Services are registered in App.cs as follows:
`public class App : MvxApplication
    {
        public override void Initialize()
        {
            CreatableTypes()
                .EndingWith("Service")
                .AsInterfaces()
                .RegisterAsLazySingleton();

            CreatableTypes()
                .EndingWith("Client")
                .AsInterfaces()
                .RegisterAsLazySingleton();

            // Register Data Managers
            CreatableTypes()
                .EndingWith(@"Manager")
                .AsInterfaces()
                .RegisterAsLazySingleton();

            //RegisterAppStart<MainViewModel>();
            RegisterCustomAppStart<AppStart>();
        }
    }`

**Comment by tkemp2290** [comment]
Sorry, the formatting seems a bit off!

**Comment by tkemp2290** [comment]
but changing DataDownloadService.cs to be this, fixes the bug but obviously doesn't provide me with the functionality I want.

`
    public class DataDownloadService : IDataDowloadService
    {
        private readonly IDataDownloadWebService _downloadService;
        private readonly ISharedPreferencesPlatformService _sharedPrefs;

        public DataDownloadService(IDataDownloadWebService downloadService, ISharedPreferencesPlatformService sharedPreferences)
        {
            _downloadService = downloadService;
            _sharedPrefs = sharedPreferences;
        }

        public async Task<DownloadedDataModel> PerformFullDownloadIfNecessary()
        {
            DownloadedDataModel downloadedData = null;
            //check last download time to see if a full download is necessary
            //var lastDownloadDate = _sharedPrefs.RetrieveDateTimePreference(Constants.LAST_DOWNLOAD_PERFORMED_DATE_KEY);
            //var differenceBetweenLastDownloadAndNow = DateTime.Now - lastDownloadDate;

            //if (differenceBetweenLastDownloadAndNow > TimeSpan.FromHours(8))
            //{
                downloadedData = await _downloadService.PerformFullDataDownload();
            //}
            return downloadedData;
        }
    }
`

**Comment by jamesmontemagno** [comment]
I see in your code that you have:

```csharp
return Preferences.Get("key", DateTime.MinValue);
```

instead it should be:
```csharp
return Preferences.Get(key, DateTime.MinValue);
```

**Comment by tkemp2290** [comment]
Hi @jamesmontemagno - I've changed that -- rather stupid typo, so thanks for pointing that out! But it hasn't changed the behaviour unfortunately. Still getting 'Handle must be valid. Parameter name: instance.' When accessing the string collection preference after accessing the datetime one.

**Comment by jamesmontemagno** [comment]
Ah, I see you have a combination of using both the Android preferences and the Xamarin.Essentials preferences. We dispose of shared preferences after using. You probably don't want to cache yours around and follow our pattern -> https://github.com/xamarin/Essentials/blob/develop/Xamarin.Essentials/Preferences/Preferences.android.cs

However if you are already using the editor, why use our preferences at all? I would just use the native android one, we call the same stuff under the hood.

**Comment by jamesmontemagno** [comment]
for date time we simply convert it to a binary long ->
```
        public static DateTime Get(string key, DateTime defaultValue, string sharedName) =>
            DateTime.FromBinary(PlatformGet<long>(key, defaultValue.ToBinary(), sharedName));

        public static void Set(string key, DateTime value, string sharedName) =>
            PlatformSet<long>(key, value.ToBinary(), sharedName);
```

**Comment by tkemp2290** [comment]
Hi @jamesmontemagno - I think I've used the combination of the two methods because Xamarin.Essentials doesn't support string collections natively and the Android SharedPreferences didn't support DateTime preferences natively. I wanted the code to be as clean as possible so I opted for the hybrid approach of using Xamarin.Essentials.Preferences and Android.SharedPreferences.

Is this not something that is supported?
